src/autoprops.rs

```diff
@@ -0,0 +1,261 @@
+                                        component_name = name;
+                                    }
+                                }
+                                return true;
+                            }
+                        }
+                    }
+                    _ => {}
+                }
+                false
+            })
+            .ok_or_else(|| {
+                syn::Error::new_spanned(
+                    sig,
+                    "could not find #[function_component] attribute in function declaration \
+                     (#[autoprops] must be placed *before* #[function_component])",
+                )
+            })?;
+
+        for input in &sig.inputs {
+            if let syn::FnArg::Typed(syn::PatType { pat, .. }) = input {
+                if let syn::Pat::Wild(wild) = pat.as_ref() {
+                    return Err(syn::Error::new_spanned(
+                        wild,
+                        "cannot use `_` as field name",
+                    ));
+                }
+            }
+        }
+
+        let properties_name = syn::Ident::new(
+            &format!("{}Props", component_name),
+            proc_macro2::Span::call_site(),
+        );
+
+        Ok(Self {
+            properties_name,
+            item_fn,
+        })
+    }
+}
+
+impl Autoprops {
+    pub fn apply_args(&mut self, args: AutopropsArgs) {
+        if let Some(name) = args.properties_name {
+            self.properties_name = name;
+        }
+    }
+
+    fn print_function_component(&self) -> proc_macro2::TokenStream {
+        let properties_name = &self.properties_name;
+        let syn::ItemFn {
+            attrs,
+            vis,
+            sig,
+            block,
+        } = &self.item_fn;
+
+        let fn_name = &sig.ident;
+        let (_impl_generics, type_generics, where_clause) = sig.generics.split_for_impl();
+        let inputs = if self.needs_a_properties_struct() {
+            // NOTE: function components currently don't accept receivers, we're just passing the
+            //       information to the next macro to fail and give its own error message
+            let receivers = sig
+                .inputs
+                .iter()
+                .filter_map(|arg| match arg {
+                    syn::FnArg::Receiver(receiver) => Some(receiver),
+                    _ => None,
+                })
+                .collect::<Vec<_>>();
+            let args = sig
+                .inputs
+                .iter()
+                .filter_map(|arg| match arg {
+                    syn::FnArg::Typed(syn::PatType { pat, .. }) => Some(quote! { #pat, }),
+                    _ => None,
+                })
+                .collect::<Vec<_>>();
+            let phantom = sig.generics.type_params().next().is_some().then(|| {
+                quote! {
+                    phantom: _,
+                }
+            });
+            quote! {
+                #(#receivers,)* #properties_name {
```

Comment:
What would the generated code be here? I'm not sure if the generated code is correct. To me, it seems that the following code generated:
```rust
self Props { ... }
```

Reply:
the whole branch is working on my big project, so i assume its correct unless i somehow dont hit this line

there is a comma not before * but inside group so id assume that if you are right it would at least be `self, Props { ... }`

Reply:
yes the `,` is inside the parenthesis expressly for that reason, there is a test that validates that &self is actually passed to function_component properly and that I get the right error message. you can see it in the stderr file:

```
error: function components can't accept a receiver
 --> tests/function_component_attr/autoprops-fail.rs:6:23
  |
6 | fn CantAcceptReceiver(&self, b: bool) -> Html {
  |                       ^^^^^
```